r/R/dplyr-summarize.R

```diff
@@ -322,15 +301,71 @@ arrow_eval_or_stop <- function(expr, mask) {
   out
 }
 
+# This function returns a list of expressions which is used to project the data
+# before an aggregation. This list includes the fields used in the aggregation
+# expressions (the "targets") and the group fields. The names of the returned
+# list are used to ensure that the projection node is wired up correctly to the
+# aggregation node.
 summarize_projection <- function(.data) {
   c(
-    map(.data$aggregations, ~ .$data),
+    unlist(unname(imap(
+      .data$aggregations,
+      ~set_names(
+        .x$data,
+        aggregate_target_names(.x$data, .y)
+      )
+    ))),
     .data$selected_columns[.data$group_by_vars]
   )
 }
 
+# This function determines what names to give to the fields used in an
+# aggregation expression (the "targets"). When an aggregate function takes 2 or
+# more fields as targets, this function gives the fields unique names by
+# appending `..1`, `..2`, etc. When an aggregate function is nullary, this
+# function returns a zero-length character vector.
+aggregate_target_names <- function(data, name) {
+  if (length(data) > 1) {
+    paste(name, seq_along(data), sep = "..")
+  } else if (length(data) > 0) {
+    name
+  } else {
+    character(0)
+  }
+}
+
+# This function returns a named list of the data types of the aggregate columns
+# returned by an aggregation
+aggregate_types <- function(.data, hash, schema = NULL) {
```

Comment:
Is schema needed as an argument here? The Expressions in $data will all bring schema with them.

Reply:
The code in `dplyr-collect.R` passes `old_schm` to this function. That's how it worked before and I didn't want to change in for fear of breaking it.

Reply:
That could predate the schema tracking in Expressions. I'd try removing it, and if the tests pass, I guess it's not needed? After all, that's what the tests are there to tell us.

Reply:
I removed it and it looks like the tests pass.

Reply:
Oops, I take that back. It looks like there is an edge case with datasets where the old schema does need to be passed, and this did cause a test to fail: https://github.com/apache/arrow/actions/runs/4147404413/jobs/7174247433#step:5:24083

Reply:
I reverted the removal and restored the schema argument here.